client/cody/src/services/LocalAppDetector.ts

```diff
@@ -0,0 +1,125 @@
+import * as vscode from 'vscode'
+
+export interface LocalProcess {
+    arch?: string
+    homeDir?: string | undefined
+    os?: string
+    isAppInstalled: boolean
+}
+
+const LOCAL_APP_LOCATIONS: { [key: string]: string[] } = {
+    // Only apply silicon is supported
+    darwin: [
+        '~/Library/Application Support/com.sourcegraph.cody',
+        '/Applications/Sourcegraph.app',
+        '/Applications/Cody.app',
+    ],
+}
+
+async function pathExists(path: string): Promise<boolean> {
+    try {
+        await vscode.workspace.fs.stat(vscode.Uri.file(path))
+        return true
+    } catch {
+        return false
+    }
+}
+
+function expandHomeDir(path: string): string {
+    if (path.startsWith('~/')) {
+        return path.replace('~', process.env.HOME || '')
+    }
+    return path
+}
+
+type OnChangeCallback = (value: boolean) => void
+
+/**
+ * Detects whether the user has the Sourcegraph app installed locally.
+ */
+export class LocalAppDetector implements vscode.Disposable {
+    private arch: string
+    private platformName: string
+    private homeDir: string | undefined
+
+    private isSupported: boolean
+    private isInstalled = false
+    private localAppMarkers: string[] | undefined
+
+    private onChange: OnChangeCallback
+    private _watchers: vscode.Disposable[] = []
+
+    constructor(options: { onChange: OnChangeCallback }) {
+        this.onChange = options.onChange
+        this.platformName = process.platform
+        this.arch = process.arch
+        this.homeDir = process.env.HOME
+        this.localAppMarkers = LOCAL_APP_LOCATIONS[this.platformName]
+        // Only Mac Silicon (M1 chip) is supported
+        this.isSupported = this.platformName === 'darwin' && this.arch === 'arm64'
+        this.start()
+    }
+
+    public async detect(): Promise<void> {
+        const startCondition = this.canStart()
+        if (!startCondition || !this.localAppMarkers) {
+            return
+        }
+        for (const marker of this.localAppMarkers) {
+            const markerExists = await pathExists(expandHomeDir(marker))
+            if (markerExists) {
+                this.fire(true)
+            }
+        }
```

Comment:
It's a bit of an eyesore but you could race the Promises:

```suggestion
if (Promise.any(this.localAppMarkers.map(marker => pathExists(expandHomeDir(marker))))) {
  this.fire(true)
}
```